File: mobile/src/main/java/com/rndapp/roostremote/models/Option.java
package com.rndapp.roostremote.models;

import com.google.gson.Gson;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.reflect.TypeToken;

import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by ell on 5/17/15.
 */
public abstract class Option {
    public abstract String getName();
    public abstract Object getValue();
    public abstract JSONObject getJsonObject(String key) throws JSONException;

    protected static void addStaticValues(JSONObject jsonObject, List<Option> staticValues) {
        if (jsonObject != null) {
            if (staticValues != null) {
                try {
                    for (Option value : staticValues){
                        jsonObject.put(value.getName(), value.getValue());
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static class OptionDeserializer implements JsonDeserializer<Option> {

        @Override
        public Option deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            JsonObject jsonObject = json.getAsJsonObject();
            return tryIntOption(jsonObject);
        }

        private Option tryIntOption(JsonObject jsonObject) {
            String name = jsonObject.get("name").getAsString();
            try{
                String intString = jsonObject.get("value").getAsString();
                int intValue = Integer.parseInt(intString);
                return new IntOption(name, intValue);
            }catch (NumberFormatException e) {
                return tryStringOption(jsonObject);
            }catch (UnsupportedOperationException ue) {
                return tryMapOption(jsonObject);
            }
        }

        private Option tryMapOption(JsonObject jsonObject) {
            String name = jsonObject.get("name").getAsString();
            Map<String, Object> map = new Gson().fromJson(jsonObject.get("value").toString(),
                    new TypeToken<HashMap<String, Object>>() {}.getType());
            if (map == null){
                return tryStringOption(jsonObject);
            }
            return new MapOption(name, map);
        }

        private Option tryStringOption(JsonObject jsonObject){
            String name = jsonObject.get("name").getAsString();
            try {
                String stringValue = jsonObject.get("value").getAsString();
                return new StringOption(name, stringValue);
            }catch (Exception e2){
                e2.printStackTrace();
                return tryBoolOption(jsonObject);
            }
        }

        private Option tryBoolOption(JsonObject jsonObject) {
            String name = jsonObject.get("name").getAsString();
            try {
                boolean boolValue = jsonObject.get("value").getAsBoolean();
                return new BoolOption(name, boolValue);
            }catch (Exception e1){
                return tryStringOption(jsonObject);
            }
        }

    }
}

File: mobile/src/main/java/com/rndapp/roostremote/models/Option.kt
package com.rndapp.roostremote.models

import com.google.gson.Gson
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.reflect.TypeToken

import org.json.JSONException
import org.json.JSONObject

import java.io.Serializable
import java.lang.reflect.Type
import java.util.HashMap

/**
 * Created by ell on 5/17/15.
 */
abstract class Option(val name: String,
                      val value: Any) : Serializable {
    @Throws(JSONException::class)
    abstract fun getJsonObject(key: String): JSONObject

    class OptionDeserializer : JsonDeserializer<Option> {

        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Option {
            val jsonObject = json.asJsonObject
            return tryIntOption(jsonObject)
        }

        private fun tryIntOption(jsonObject: JsonObject): Option {
            val name = jsonObject.get("name").asString
            try {
                val intString = jsonObject.get("value").asString
                val intValue = Integer.parseInt(intString)
                return IntOption(name, intValue)
            } catch (e: NumberFormatException) {
                return tryStringOption(jsonObject)
            } catch (ue: UnsupportedOperationException) {
                return tryMapOption(jsonObject)
            }

        }

        private fun tryMapOption(jsonObject: JsonObject): Option {
            val name = jsonObject.get("name").asString
            val map = Gson().fromJson<Map<String, Any>>(jsonObject.get("value").toString(),
                    object : TypeToken<HashMap<String, Any>>() {

                    }.type) ?: return tryStringOption(jsonObject)
            return MapOption(name, map)
        }

        private fun tryStringOption(jsonObject: JsonObject): Option {
            val name = jsonObject.get("name").asString
            try {
                val stringValue = jsonObject.get("value").asString
                if (stringValue == "true" || stringValue == "false") {
                    val boolValue = jsonObject.get("value").asBoolean
                    return BoolOption(name, boolValue)
                }
                return StringOption(name, stringValue)
            } catch (e2: Exception) {
                e2.printStackTrace()
                return tryBoolOption(jsonObject)
            }

        }

        private fun tryBoolOption(jsonObject: JsonObject): Option {
            val name = jsonObject.get("name").asString
            try {
                val boolValue = jsonObject.get("value").asBoolean
                return BoolOption(name, boolValue)
            } catch (e1: Exception) {
                return tryStringOption(jsonObject)
            }

        }

    }

    companion object {
        fun addStaticValues(jsonObject: JSONObject?, staticValues: List<Option>?) {
            if (jsonObject != null) {
                if (staticValues != null) {
                    try {
                        for (value in staticValues) {
                            jsonObject.put(value.name, value.value)
                        }
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }

                }
            }
        }
    }
}
